packages/providers/src/transaction-response/transaction-response.ts

```diff
@@ -199,11 +213,24 @@ export class TransactionResponse {
     const result = await this.waitForResult<TTransactionType>(contractsAbiMap);
 
     if (result.isStatusFailure) {
-      throw new Error(
+      throw new FuelError(
+        ErrorCode.TRANSACTION_FAILED,
         `Transaction failed: ${(<FailureStatus>result.gqlTransaction.status).reason}`
       );
     }
 
     return result;
   }
+
+  private async sleepBasedOnAttempts(attempts: number): Promise<void> {
+    // This code implements a similar approach from the fuel-core await_transaction_commit
+    // https://github.com/FuelLabs/fuel-core/blob/cb37f9ce9a81e033bde0dc43f91494bc3974fb1b/fuel-client/src/client.rs#L356
+    // double the interval duration on each attempt until max is reached
+    //
+    // This can wait forever, it would be great to implement a max timeout here, but it would require
+    // improve request handler as response Error not mean that the tx fail.
```

Comment:
The code here is simple; the previous comments felt redundant to me.

Except for the `maxTimeout` idea that I turned into a TODO comment.


```suggestion
  /*
    TODO: Consider adding `maxTimeout` or `maxAttempts` parameter.

    The aim is to avoid perpetual execution; when the limit
    is reached, we can throw accordingly.
  */
  private async sleepBasedOnAttempts(attempts: number): Promise<void> {
```